### ISSUE TITLE
[Bug]: React 18 - `hydrateRoot` + `StrictMode` breaks with invariant error

### ISSUE THREAD

**Issue by CGamesPlay** [airjp73/remix-validated-form#109]
### Which packages are impacted?

- [X] `remix-validated-form`
- [ ] `@remix-validated-form/with-zod`
- [ ] `@remix-validated-form/with-yup`
- [ ] `zod-form-data`

### What version of these packages are you using?

- `remix-validated-form`: `^4.4.0`
-  `@remix-validated-form/with-zod`: `^2.0.1`

### Please provide a link to a minimal reproduction of the issue.

https://github.com/airjp73/remix-validated-form/blob/main/packages/remix-validated-form/src/internal/state/createFormStore.ts#L105

### Steps to Reproduce the Bug or Issue

Apologies that I don't have a 1-click repro. I'm pretty sure I'm doing exactly what's shown in the [demo](https://www.remix-validated-form.io/), although since the demo code doesn't ever show the full code I can't be totally sure. However, I can tell you that the code works in `4.3.1` and fails in `4.4.0`, and when I used the debugger to investigate, I can see that `setFormElement` is being called with the proper element, but when `validateField` is called, `get().formElement` is null. So it seems like the draft is never being saved or something like that.

### Expected behavior

The invariant should not be varied.

### Platform

- OS: macOS
- Browser: Chrome
- Version: All

**Comment by airjp73** [comment]
A reproduction would definitely be helpful here. The demo on the site always uses the latest version of the library since it's a monorepo, so there must be something different going on in your case.

Can you describe a bit more what you have to do to break it? Does it just break on mount or when you blur an input or something else?

If it helps, the `FormInput` in the demo is [here](https://github.com/airjp73/remix-validated-form/blob/main/apps/docs/app/components/FormInput.tsx), and all the other components imported in the demo are in the same folder.

**Comment by CGamesPlay** [comment]
Turns out it's caused by `React.StrictMode`. The [attached zip file](https://github.com/airjp73/remix-validated-form/files/8667966/repo.zip) includes a repository with a minimal reproduction. Extract, run `yarn; remix dev`, open the page, click in the input, and blur it. The invariant violation will appear in the console. If you remove `React.StrictMode` from `entry.client.tsx`, the problem is bypassed. [Documentation about React Strict Mode](https://reactjs.org/docs/strict-mode.html). In particular, as documented in [Detecting unexpected side effects](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects) the "double-call of render" is likely the culprit here. It's important to note that this likely does describe a real bug, like incorrectly utilizing a reference that isn't tracked through standard hooks.

For easy reference, the dependencies listed in the test repo are:

```json
{
  "dependencies": {
    "@remix-run/node": "^1.4.3",
    "@remix-run/react": "^1.4.3",
    "@remix-run/serve": "^1.4.3",
    "@remix-run/server-runtime": "^1.0.0",
    "@remix-validated-form/with-zod": "^2.0.1",
    "react": "^18.1.0",
    "react-dom": "^18.1.0",
    "remix-validated-form": "^4.4.0",
    "zod": "^3.15.1"
  },
  "devDependencies": {
    "@remix-run/dev": "^1.4.3",
    "@types/react": "^18.0.8",
    "@types/react-dom": "^18.0.3",
    "typescript": "^4.5.5"
  }
}
```

**Comment by marcinczenko** [comment]
I experience the problem only on Safari: https://github.com/marcinczenko/remix-starter-with-tailwindcss.

**Comment by airjp73** [comment]
I have a fix locally, but now have a broken test. Will probably be ready this week sometime.

**Comment by airjp73** [comment]
I've published the fix as [4.5.0-beta.0](https://www.npmjs.com/package/remix-validated-form/v/4.5.0-beta.0). It's a big enough refactor that feels weird to release as a patch. I have a few other small features I want to squeeze into 4.5 so I'll do a full release once those are done.

**Comment by airjp73** [comment]
Fix released in [4.5.0](https://github.com/airjp73/remix-validated-form/releases/tag/remix-validated-form-v4.5.0)
